File: icode/ch09_shortest_path/dijkstra.js
console.time();

const [n, m] = [6, 11];
const start = 1;
const arr = [
  [1, 2, 2],
  [1, 3, 5],
  [1, 4, 1],
  [2, 3, 3],
  [2, 4, 2],
  [3, 2, 3],
  [3, 6, 5],
  [4, 3, 3],
  [4, 5, 1],
  [5, 3, 1],
  [5, 6, 2],
];

const visited = Array(n + 1).fill(false);
const dist = Array(n + 1).fill(Infinity);
const graph = Array.from(Array(n + 1), () => []);

for (const [a, b, c] of arr) {
  graph[a].push([b, c]);
}

// 방문하지 않은 노드 중, 최단 거리가 가장 짧은 노드의 인덱스 반환
function getSmallNode() {
  let min = Infinity;
  let idx = 0;

  for (let i = 1; i < n + 1; i++) {
    if (dist[i] < min && !visited[i]) {
      min = dist[i];
      idx = i;
    }
  }
  return idx;
}

// 시작 노드 초기화
dist[start] = 0;
visited[start] = true;
for (const [idx, cost] of graph[start]) {
  dist[idx] = cost;
}

// 제외한 n - 1개의 노드에 대해 반복
for (let i = 0; i < n - 1; i++) {
  const now = getSmallNode();
  visited[now] = true;

  for (const [idx, cost] of graph[now]) {
    const newCost = dist[now] + cost;
    if (newCost < dist[idx]) {
      dist[idx] = newCost;
    }
  }
}

for (let i = 1; i <= n; i++) {
  if (dist[i] === Infinity) {
    console.log("INFINITY");
  } else {
    console.log(dist[i]);
  }
}

console.timeEnd();

File: source/shortest-path/dijkstra.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

const int INF = 1e9;

int n, m, start;
vector<pair<int, int>> graph[100001];
int d[100001];


/**
 * ### 최소의 누적거리를 선택하며 갱신
 *
 * pq
 * - param1: 해당 노드까지의 누적 거리
 * - param2: 탐색할 노드의 번호
 *
 * d
 * - 해당 노드(인덱스)까지의 최소 거리를 나타내는 일차원 배열
 */
void dijkstra(int start) {
    priority_queue<pair<int, int> > pq;

    d[start] = 0;
    pq.push({ 0, start });

    while (!pq.empty()) {
        int dist = -pq.top().first;
        int now = pq.top().second;
        pq.pop();

        if (d[now] < dist) continue;

        for (auto& pair : graph[now]) {
            int cost = dist + pair.second;
            int next = pair.first;
            if (cost < d[next]) {
                d[next] = cost;
                pq.push({ -cost, next });
            }
        }
    }
}

int main(void) {
    cin >> n >> m >> start;

    for (int i = 0; i < m; i++) {
        int a, b, c;
        cin >> a >> b >> c;
        graph[a].push_back({ b, c });
    }

    fill(d, d + 100001, INF);
    dijkstra(start);

    for (int i = 1; i <= n; i++) {
        if (d[i] == INF) cout << "INFINITY" << '\n';
        else cout << d[i] << '\n';
    }
}
